### PR TITLE
feat: Add Nix Builder + Add remote runner example plan

### PR DESCRIPTION
**Pull request by MarcoPolo**
This PR does a couple things:

1. Introduces a Nix builder that lets us build binaries, containers, buildpacks and OSs for multiple architectures. (you can actually define what the OS looks like and deploy it as part of the test!)
  a. Adds an integration test (that will fail until we add Nix to Circle CI)
2. Tweaks `local:exec` to allow you to run any executable, not just Go binaries.
3. Adds a new example plan that demonstrates how to support remote runners with `ssh` and the Nix builder. It leverages Nix to build for different architectures and efficiently/correctly deploy the application to the remote machine – even if the application is dynamically linked (caveat that the remote machine must have Nix installed).
  a. This plan tests a simple server/client ping across machines.
  

In the long term we'll probably want some sort of first class support for remote runners in testground, but I think this is a usable solution in the short term that will let us set up some reproducible benchmark environments. But I may be missing something in my understanding of how Testground works, so apologies if that's the case.

r? @laurentsenta

### CODE REVIEW

**Comment by p-shahi** [comment]
@laurentsenta and I discussed this PR today and what it would enable for the libp2p maintainers.

We specifically [talked about this proposal](https://github.com/libp2p/go-libp2p/pull/1804#issuecomment-1271995750)
>> That being said. I can add to the benchmark milestone and move F.1 to Q4 (or early Q1): wdyt? F.1 - Benchmarking using nix-builders (dependency: remote machines need Nix installed) F.2 - Benchmarking using first class support for remote runners (using remote:exec) in Testground

>This sound good. I think F.1 will give us >80% of the benefit of F.2. And also inform what F.2 should look like. Also worth noting that we can do F.1 _today_. We aren't blocked on anything (except time and prioritization).

For that reason, this PR has been prioritized and added to both the test-plans and go-libp2p roadmaps (for Q2 November) to bootstrap the go-libp2p benchmarking efforts.
It will also enable running [js-libp2p-transfer-performance](https://github.com/ipfs-shipyard/js-libp2p-transfer-performance) as a test-plan and CI job to benchmark transfer times across releases (to catch issues like [#1342](https://github.com/libp2p/js-libp2p/issues/1342))

The questions we had were: 
 - If this PR achieves 80% of what we need, what is the remaining 20% that needs to be addressed afterwards?[^1]

A question Laurent had with regard to the Testground team's project prioritization was:
 - Is this PR enough to unblock the libp2p teams, and allow the Testground maintainers to prioritize the set of network simulation features (used currently in CI) to do RTT testing as a part of [multi-dimenstional testing](https://github.com/testground/testground/issues/1392) and deprioritize https://github.com/testground/testground/issues/1392 [^2]

cc: @BigLep @tinytb 

[^1]: My thoughts were, I think we will likely have a better idea of the "20% gap" after merging and using this in Testground. It largely accomplishes what `remote:exec` is trying to do ["we build a binary, scp to the remote machine, and start the binary there"](https://www.notion.so/pl-strflt/Remote-Runners-c4ad4886c4294fb6a6f8afd9c0c5b73c#cd6d237911d148f18fb5620e5ce1d04f)
[^2]: I think yes it's enough to allow the Testground team to focus on the network simulation features.

File: plans/remote-runner/README.md
Comment (p-shahi) [inline plans/remote-runner/README.md:13]:
```suggestion
2. Using a `local:exec` runner to run a shell script that copies the built
```
Numbering is wrong.

Also, would it be better define a new `remote:exec` runner?

**Comment by MarcoPolo** [comment]
@galargh friendly ping on this

**Comment by galargh** [comment]
> @galargh friendly ping on this

It's on my TODO list for this week. I'll have it reviewed on Wednesday the latest - hopefully, tomorrow. Sorry for the delay.

File: plans/remote-runner/README.md
Comment (galargh) [reply plans/remote-runner/README.md:13]: If you keep using `1.`s for list items then GitHub Markdown creates a numbered list starting from `1`. That's where the numbering comes from.

As for the name, `local:exec` is an existing construct. Also, the name is fitting here because the shell script that performs the copy is executed locally.

File: pkg/build/nix_generic.go
Comment (galargh) [inline pkg/build/nix_generic.go:29]: Am I right in saying it builds whatever `flake.nix` specifies so it doesn't necessarily have to be Go?

File: pkg/build/nix_generic.go
Comment (galargh) [inline pkg/build/nix_generic.go:20]: Do we care about distinguishing between binary/container artifacts? Should we flag it somehow?

File: pkg/build/nix_generic.go
Comment (galargh) [inline pkg/build/nix_generic.go:58]: What's this for?

File: pkg/runner/local_exec.go
Comment (galargh) [inline pkg/runner/local_exec.go:51]: Just to confirm I understand this. We need this for nix. It outputs an artifact. We pass the artifact path to local exec runner. We want to execute a specific binary within that artifact. Is it correct?

**Review by galargh — APPROVED** [review]
Looks good :) We'll probably want to update docs at some point but I think we can hold of for a bit and see how we end up using it in practice.

I left a few questions here and there, mostly to make sure I understand it correctly.

🚀

**Comment by galargh** [comment]
@laurentsenta What was the trick for `go mod tidy` not to look into directories with different go versions? It seems that's the problem here.

**Comment by laurentsenta** [comment]
@galargh I used a `_` prefix,

https://pkg.go.dev/cmd/go#hdr-Package_lists_and_patterns

> Directory and file names that begin with "." or "_" are ignored by the go tool, as are directories named "testdata".

@MarcoPolo is this still priority if we plan on moving the building out of testground in `libp2p/test-plans`?

**Comment by MarcoPolo** [comment]
Closing this. I think we want to get out of the building game. And I think we need to make testground simpler to enable remote workers properly.

**Pull request closed without merging**
